nextcord/cooldowns/cooldown.py

```diff
@@ -0,0 +1,253 @@
+
+    return decorator
+
+
+class _CooldownTimesPer:
+    # Essentially TimesPer but modified
+    # to throw Exceptions instead of queue
+    def __init__(
+        self,
+        limit: int,
+        per: float,
+        _cooldown: Cooldown,
+    ) -> None:
+        self.limit: int = limit
+        self.per: float = per
+        self._cooldown: Cooldown = _cooldown
+        self.current: int = limit
+        self.loop: AbstractEventLoop = get_event_loop()
+        self.pending_reset: bool = False
+
+    async def __aenter__(self) -> "_CooldownTimesPer":
+        if self.current == 0:
+            raise CallableOnCooldown(self._cooldown.func, self._cooldown, self.per)
+
+        self.current -= 1
+
+        if not self.pending_reset:
+            self.pending_reset = True
+            self.loop.call_later(self.per, self.reset)
+
+        return self
+
+    async def __aexit__(self, *_) -> None:
+        ...
+
+    def reset(self):
+        self.current = self.limit
+
+
+class Cooldown:
+    """Represents a cooldown for any given :type:`Callable`."""
+
+    def __init__(
+        self,
+        limit: int,
+        per: float,
+        bucket: Optional[Bucket] = None,
+        func: Optional[Callable] = None,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        limit: int
+            How many call's can be made in the time
+            period specified by ``per``
+        per: float
+            The time period related to ``limit``
+        bucket: Optional[Bucket]
+            The :class:`Bucket` implementation to use
+            as a bucket to separate cooldown buckets.
+
+            Defaults to :class:`CooldownBucket.all`
+        func: Optional[Callable]
+            The function this cooldown is attached to
+        """
+        bucket = bucket or CooldownBucket.all
+        # See TimesPer for inspo, however it doesnt
+        # use that due to the required changes and flexibility
+        self.limit: int = limit
+        self.per: float = per
+
+        self._func: Optional[Callable] = func
+        self._bucket: Bucket = bucket
+        self.loop: AbstractEventLoop = get_event_loop()
+        self.pending_reset: bool = False
+        self.last_reset_at: Optional[float] = None
+        self._last_bucket: Optional[_HashableArguments] = None
+
+        self._cache: dict[_HashableArguments, _CooldownTimesPer] = {}
+
+    async def __aenter__(self) -> "Cooldown":
+        bucket: _CooldownTimesPer = self._get_cooldown_for_bucket(self._last_bucket)
+        async with bucket:
+            return self
+
+    async def __aexit__(self, *_) -> None:
+        ...
+
+    def __call__(self, *args, **kwargs):
+        self._last_bucket = self.get_bucket(*args, **kwargs)
+        return self
+
+    def _get_cooldown_for_bucket(self, bucket: _HashableArguments):
+        try:
+            return self._cache[bucket]
+        except KeyError:
+            _bucket = _CooldownTimesPer(self.limit, self.per, self)
```

Comment:
This will never be cleared? Also maybe use something like TTLCache (or a custom version which resets the cooldown after use?)